File: eagle-core/eagle-data-process/eagle-stream-process-api/src/main/scala/org/apache/eagle/datastream/storm/AbstractStreamBolt.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.eagle.datastream.storm

import java.util

import backtype.storm.task.{OutputCollector, TopologyContext}
import backtype.storm.topology.OutputFieldsDeclarer
import backtype.storm.topology.base.BaseRichBolt
import backtype.storm.tuple.{Fields, Tuple}
import org.apache.eagle.datastream.core.StreamInfo
import org.apache.eagle.datastream.utils.NameConstants
import org.slf4j.LoggerFactory

/**
 *
 * @param fieldsNum zero-fieldsNum may means something different
 * @param ack
 * @param streamInfo
 * @tparam T
 */
abstract class AbstractStreamBolt[T](val fieldsNum:Int=0, val ack:Boolean = true)(implicit streamInfo:StreamInfo) extends BaseRichBolt{
  private var _collector: OutputCollector = null
  private val LOG = LoggerFactory.getLogger(classOf[AbstractStreamBolt[T]])

  /**
   * If outKeyed then
   *  Fields = ("key","value"]
   * elsif num > 0
   *  Fields = ["f0","f1",..,"fn"]
   * elsif num == 0
   *  Fields = ["f0"]
   * end
   *
   * @param declarer
   */
  override def declareOutputFields(declarer: OutputFieldsDeclarer): Unit = {
    if(streamInfo.outKeyed) {
      declarer.declare(new Fields(NameConstants.FIELD_KEY,NameConstants.FIELD_VALUE))
    }else{
      if(fieldsNum > 0) {
        val fields = new util.ArrayList[String]()
        var i: Int = 0
        while (i < fieldsNum) {
          fields.add(NameConstants.FIELD_PREFIX + i)
          i += 1
        }
        declarer.declare(new Fields(fields))
      }else if(fieldsNum == 0){
        declarer.declare(new Fields(NameConstants.FIELD_PREFIX + 0))
      }
    }
  }

  def emit(values:util.List[AnyRef])(implicit input:Tuple){
    if (streamInfo.outKeyed) {
      _collector.emit(input, util.Arrays.asList(streamInfo.keySelector.key(values).asInstanceOf[AnyRef], values))
    } else {
      _collector.emit(input, values)
    }
  }

  def emit(value:Any)(implicit input:Tuple){
    if(streamInfo.outKeyed) {
      _collector.emit(input, util.Arrays.asList(streamInfo.keySelector.key(value).asInstanceOf[AnyRef],value.asInstanceOf[AnyRef]))
    }else{
      _collector.emit(input,util.Arrays.asList(value.asInstanceOf[AnyRef]))
    }
  }

  override def execute(input: Tuple): Unit = {
    try {
      implicit val _input = input
      if (streamInfo.inKeyed) {
        val key = input.getValueByField(NameConstants.FIELD_KEY)
        val value = input.getValueByField(NameConstants.FIELD_VALUE).asInstanceOf[T]
        onKeyValue(key, value)
      } else {
        onValues(input.getValues)
      }
      if(ack) _collector.ack(input)
    }catch {
      case t: Throwable => {
        LOG.error(s"Got exception when processing $input",t)
        _collector.fail(input)
      }
    }
  }

  /**
   * Handle keyed stream value
   */
  def onKeyValue(key:Any,value:T)(implicit input:Tuple)

  /**
   * Handle general stream values list
   *
   * @param values
   */
  def onValues(values:util.List[AnyRef])(implicit input:Tuple)

  override def prepare(stormConf: util.Map[_, _], context: TopologyContext, collector: OutputCollector): Unit = {
    _collector = collector
    streamInfo.reinit()
  }
}

File: eagle-core/eagle-alert-parent/eagle-alert/alert-engine/src/main/java/org/apache/eagle/alert/engine/runner/AbstractStreamBolt.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.eagle.alert.engine.runner;

import org.apache.eagle.alert.engine.StreamContext;
import org.apache.eagle.alert.engine.coordinator.IMetadataChangeNotifyService;
import org.apache.eagle.alert.engine.coordinator.StreamDefinition;
import org.apache.eagle.alert.engine.coordinator.StreamNotDefinedException;
import org.apache.eagle.alert.engine.model.PartitionedEvent;
import org.apache.eagle.alert.engine.serialization.PartitionedEventSerializer;
import org.apache.eagle.alert.engine.serialization.SerializationMetadataProvider;
import org.apache.eagle.alert.engine.serialization.Serializers;
import org.apache.eagle.alert.utils.AlertConstants;
import org.apache.storm.task.OutputCollector;
import org.apache.storm.task.TopologyContext;
import org.apache.storm.topology.OutputFieldsDeclarer;
import org.apache.storm.topology.base.BaseRichBolt;
import org.apache.storm.tuple.Fields;
import com.google.common.base.Preconditions;
import com.typesafe.config.Config;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@SuppressWarnings( {"rawtypes", "serial"})
public abstract class AbstractStreamBolt extends BaseRichBolt implements SerializationMetadataProvider {
    private static final Logger LOG = LoggerFactory.getLogger(AbstractStreamBolt.class);
    private IMetadataChangeNotifyService changeNotifyService;

    public Config getConfig() {
        return config;
    }

    private Config config;
    private List<String> outputStreamIds;
    protected OutputCollector collector;
    protected Map stormConf;

    private String boltId;
    protected PartitionedEventSerializer serializer;
    protected volatile Map<String, StreamDefinition> sdf = new HashMap<String, StreamDefinition>();
    protected volatile String specVersion = "Not Initialized";
    protected volatile boolean specVersionOutofdate = false;
    protected StreamContext streamContext;

    public AbstractStreamBolt(String boltId, IMetadataChangeNotifyService changeNotifyService, Config config) {
        this.boltId = boltId;
        this.changeNotifyService = changeNotifyService;
        this.config = config;
    }

    public void declareOutputStreams(List<String> outputStreamIds) {
        this.outputStreamIds = outputStreamIds;
    }

    protected List<String> getOutputStreamIds() {
        return this.outputStreamIds;
    }

    @Override
    public void prepare(Map stormConf, TopologyContext context, OutputCollector collector) {
        Preconditions.checkNotNull(this.changeNotifyService, "IMetadataChangeNotifyService is not set yet");
        this.stormConf = stormConf;
        this.collector = collector;
        this.serializer = Serializers.newPartitionedEventSerializer(this);
        internalPrepare(collector, this.changeNotifyService, this.config, context);
        try {
            this.changeNotifyService.activateFetchMetaData();
        } catch (Exception e) {
            LOG.warn(e.getMessage(), e);
        }
    }


    protected PartitionedEvent deserialize(Object object) throws IOException {
        // byte[] in higher priority
        if (object instanceof byte[]) {
            return serializer.deserialize((byte[]) object);
        } else if (object instanceof PartitionedEvent) {
            return (PartitionedEvent) object;
        } else {
            throw new IllegalStateException(String.format("Unsupported event class '%s', expect byte array or PartitionedEvent!", object == null ? null : object.getClass().getCanonicalName()));
        }
    }

    /**
     * subclass should implement more initialization for example.
     * 1) register metadata change
     * 2) init stream context
     *
     * @param collector
     * @param metadataManager
     * @param config
     * @param context
     */
    public abstract void internalPrepare(
        OutputCollector collector,
        IMetadataChangeNotifyService metadataManager,
        Config config, TopologyContext context);

    @Override
    public void cleanup() {
        super.cleanup();
    }

    @Override
    public void declareOutputFields(OutputFieldsDeclarer declarer) {
        if (this.outputStreamIds != null) {
            LOG.info("declare streams: {} ", outputStreamIds);
            for (String streamId : this.outputStreamIds) {
                declarer.declareStream(streamId, new Fields(AlertConstants.FIELD_0));
            }
        } else {
            declarer.declare(new Fields(AlertConstants.FIELD_0));
        }
    }

    @Override
    public StreamDefinition getStreamDefinition(String streamId) throws StreamNotDefinedException {
        if (sdf.containsKey(streamId)) {
            return sdf.get(streamId);
        } else {
            throw new StreamNotDefinedException(streamId, specVersion);
        }
    }

    public String getBoltId() {
        return boltId;
    }
}